### DISCUSSION TITLE
How do I get started with this after install?

### DISCUSSION THREAD

**Discussion by JimKillock** [Automattic/wordpress-activitypub#155] [Q&A]
I've installed the plug in, and BuddyPress in case it is needed. I can't see, however, any output from the plug in as such. 

Do I need to take some extra steps to ensure that information about how to follow etc is displayed? Or has something gone wrong with my install (on WP's hosted site)?

**Comment by Cambridgeport90** [comment]
You shouldn't have to do anything (though BuddyPress is also a fantastic piece of software, but not needed for this plugin to work.) The dinformation which you seek is found under your profile; under followers (Fediverse). You can see those who follow you; notice that you will immediately be discovered; look at it on another fediverse instance to which you have access.

**Reply by JimKillock** [reply]
Thanks - perhaps I just don’t know Wordpress well enough. My "user profile" seems to be a generic Wordpress.com page; does the plug in assume users are local only? Or is there a way to create or view a "local user" profile page if using the hosted version of Wordpress?

**Reply by pfefferle** [reply]
I have not tested the plugin with WordPress.com hosting yet to be honest! Can you send me your author URL, then I will check if I can find an issue.

**Reply by JimKillock** [reply]
Thanks! Hwre is the [website](https://openlatin.org); here is an [author page](https://openlatin.org/author/jimkillock/) and here is my user page via [Wordpress forums](https://wordpress.com/forums/users/jimkillock/). 

I can't seem to locate an actual public profile on Wordpress, tho I imagine there is one.

**Reply by pfefferle** [reply]
It seems to work fine for me. I can follow your profile and see one of your posts :)

![image](https://user-images.githubusercontent.com/2373/173894741-4f4fd745-f0b3-48d7-ad60-bb71236e51c1.png)

![image](https://user-images.githubusercontent.com/2373/173894781-a8382e5a-3bf7-4f15-a5d6-33f6a30ef9c2.png)

![image](https://user-images.githubusercontent.com/2373/173894803-117774b2-3fbc-4b85-a1b6-9845394a631e.png)

**Reply by JimKillock** [reply]
Thanks! That's great. But where does someone external find that this facility exists? Do I add it manually, or should it appear on a public page automatically somwhere?

**Reply by pfefferle** [reply]
You have to tell your readers. There is no auto generated info/page (yet).

**Reply by JimKillock** [reply]
Thanks very much for all your help :) Now to try to get some new people to use your plugin …

**Reply by stevenjaycohen** [reply]
Also, try adding your site to some of the emerging Fediverse directories to increase discoverability!
[https://fediverse.info/explore/people#howItWorks](url)

**Comment by AndySoc1al** [comment]
I'm running WordPress on my own domain, with the URLs being at the root level, no subdirectories. I've installed WebFinger and ActivityPub, I see my profile identifier says I should search for `@andysocial@andysocial.com` on Mastodon. I search that and get 0 results.

I'm stumped.

**Reply by mediaformat** [reply]
A quick debug tool is to test the [webfinger](https://webfinger.net/), yours is returning `403 Forbidden`

**Reply by pfefferle** [reply]
as @mediaformat says: something seems to block the `.well-known` URI: https://andysocial.com/.well-known/webfinger?resource=acct%3Aandysocial%40andysocial.com. This could be a setup problem or your hoster blocks it (this is often because of lets-encrypt)

**Reply by timnolte** [reply]
> as @mediaformat says: something seems to block the `.well-known` URI: https://andysocial.com/.well-known/webfinger?resource=acct%3Aandysocial%40andysocial.com. This could be a setup problem or your hoster blocks it (this is often because of lets-encrypt)

Yeah, I'd bet that it's partly caused by Let's Encrypt as I had that problem in the past. I have a different hosting environment and my Let's Encrypt setup isn't creating that directory anymore so it's not a conflict. You may have to setup an `.htaccess` rule that rewrites the webfinger request to route through WordPress.

**Reply by ricmac** [reply]
@timnolte @pfefferle I'm a bit unsure what you mean by "route through WordPress." Can you please clarify? Is there an example of an .htaccess rule for this issue that non-devs can copy and paste? Seems like this has been a known issue for sometime, ref https://github.com/pfefferle/wordpress-activitypub/issues/5#issuecomment-522057022.

**Reply by timnolte** [reply]
> @timnolte @pfefferle I'm a bit unsure what you mean by "route through WordPress." Can you please clarify? Is there an example of an .htaccess rule for this issue that non-devs can copy and paste? Seems like this has been a known issue for sometime, ref https://github.com/pfefferle/wordpress-activitypub/issues/5#issuecomment-522057022.

@ricmac I don't have a computer available ATM as I'm out of town, I can try to do some digging/testing to confirm something that might work. I was trying to see if there were any examples I could dig up but haven't found anything.

**Reply by AndySoc1al** [reply]
> > as @mediaformat says: something seems to block the `.well-known` URI: https://andysocial.com/.well-known/webfinger?resource=acct%3Aandysocial%40andysocial.com. This could be a setup problem or your hoster blocks it (this is often because of lets-encrypt)
> 
> Yeah, I'd bet that it's partly caused by Let's Encrypt as I had that problem in the past. I have a different hosting environment and my Let's Encrypt setup isn't creating that directory anymore so it's not a conflict. You may have to setup an `.htaccess` rule that rewrites the webfinger request to route through WordPress.

I haven't messed about with .htaccess in a few years, so any leads on how to do that would be great. I'm on DreamHost, and am using the Let's Encrypt service to maintain SSL certs. But, "route through WordPress" is a series of words that I can't parse into action. :-)

**Reply by timnolte** [reply]
@AndySoc1al I'm just now back from a family vacation, and back to a computer. Let me see if I can put something together for you to test out.

**Reply by timnolte** [reply]
@AndySoc1al I thought for sure that this has been covered in the past. Here is an older issue that talks about some of this issue and possible solutions. Do you know if Dreamhost is using Apache or Nginx? https://github.com/pfefferle/wordpress-activitypub/issues/8#issuecomment-1138278289

**Reply by ricmac** [reply]
@timnolte thanks so much for looking into this. It sounds like @AndySoc1al and I have the same issue. I am on a Hostinger server, and they apparently use LiteSpeed: https://www.coralnodes.com/hostinger-review/#h3_3

Ref: "While the majority of hosts provide Apache or Nginx servers, Hostinger uses LiteSpeed"

**Reply by AndySoc1al** [reply]
@timnolte Dreamhost uses Apache (currently 2.4.29 on Ubuntu). When I look at my WordPress problem menu, I see the following for this specific plugin: 

> Your WebFinger endpoint https://andysocial.com/.well-known/webfinger?resource=acct:andysocial@andysocial.com does not return valid JSON for application/jrd+json.

**Reply by Bymiki** [reply]
I have the same problem.

Site Health message:
"WebFinger endpoint is not accessible
Your WebFinger endpoint https://midomain.com/.well-known/webfinger?resource=acct:bymiki@mydomain.com does not return valid JSON for application/jrd+json."

I have tried on a fresh installation, without other plugins and with the default WP theme. I have tried with and without LetsEncrypt. I tried also with and without NodeInfo and WebFinger plugins. I get same result all the time.

This is the error log of my server:

2022-11-12 08:48:47 UTC [nginx][error] 2283#0: *34722024 openat() "/home/u1702-w4en9qhtzj5x/www/mydomain.com/public_html/.well-known/webfinger" failed (2: No such file or directory), client: 34.90.88.177, server: fediverso.eu, request: "GET /.well-known/webfinger?resource=acct:bymiki@mydomain.com HTTP/1.1", host: "mydomain.com", referrer: "https://mydomain.com/.well-known/webfinger?resource=acct:bymiki@mydomain.com"
‮
2022-11-12 07:07:24 UTC [nginx][error] 90102#0: *34568592 openat() "/home/u1702-w4en9qhtzj5x/www/mydomain.com/public_html/.well-known/security.txt" failed (2: No such file or directory), client: 188.165.87.107, server: mydomain.com, request: "GET /.well-known/security.txt HTTP/1.1", host: "www.mydomain.com"

**Reply by joshisanonymous** [reply]
I think I'm having a similar issue. I tried adding the following line to in a couple places in my `.htaccess` file:

`RedirectMatch "^\/\.well-known(.*)$" "\/blog\/\.well-known$1"`

But I a 404 error when I check webfinger.net. I also have the following in that file:

`#BEGIN Really Simple SSL LETS ENCRYPT`
`RewriteRule ^.well-known/(.*)$ - [L]`
`#END Really Simple SSL LETS ENCRYPT`

So I'm assuming this let's encrypt thing that's being discussed is impacting me, as well, but I honestly have no idea how anything works in an `.htaccess` file.

(My WordPress install is indeed in a subdirectory called `/blog/`. The site is [joshmcneill.com](https://joshmcneill.com) if that helps.)

**Reply by timnolte** [reply]
You have to restrict your redirect match to only the webfinger endpoints and then those need to redirected to WordPress. I'll have to try and setup a test site in order to figure out an example configuration.

**Reply by ricmac** [reply]
@timnolte I tried several specific redirects, but I think my web host (Hostinger) disallows any messing with the well-known folder.

**Reply by joshisanonymous** [reply]
Ok, I finally got the webfinger to work as well as being able to search for my blog from Mastodon by adding the following to the `.htaccess` in my root folder:

`RedirectMatch "\/\.well-known\/webfinger(.*)$" "\/wp-json\/activitypub\/1\.0\/webfinger$1"`

The redirect that I tried in [the post above](https://github.com/pfefferle/wordpress-activitypub/discussions/155#discussioncomment-4191499) came from this repository's `README.md`, but seemed to originate from [here](https://wordpress.org/support/topic/activitypub-does-not-work/) where pfefferle had already made it clear that such a redirect was not a good idea, so I'm not sure why it's in the `README.md`.

What still doesn't seem to work still: I can't follow my blog from Mastodon. It goes in as a pending follow request.

**Reply by AndySoc1al** [reply]
> Ok, I finally got the webfinger to work as well as being able to search for my blog from Mastodon by adding the following to the `.htaccess` in my root folder:
> 
> `RedirectMatch "\/\.well-known\/webfinger(.*)$" "\/wp-json\/activitypub\/1\.0\/webfinger$1"`
> 
> The redirect that I tried in [the post above](https://github.com/pfefferle/wordpress-activitypub/discussions/155#discussioncomment-4191499) came from this repository's `README.md`, but seemed to originate from [here](https://wordpress.org/support/topic/activitypub-does-not-work/) where pfefferle had already made it clear that such a redirect was not a good idea, so I'm not sure why it's in the `README.md`.
> 
> What still doesn't seem to work still: I can't follow my blog from Mastodon. It goes in as a pending follow request.

That did nothing for me, even after forcing a cache purge, and testing from a command line curl.

**Reply by mateuswetah** [reply]
@ricmac have you had any luck with Hostinger? Been trying several things here, nothing works. Adding my own `.well-known/webfinger` does removes warnings but I still cannot find myself in Mastodon.

**Reply by pfefferle** [reply]
@mateuswetah can you provide us your blogs domain? maybe I can have a look?

**Reply by mateuswetah** [reply]
Hey @pfefferle sure, I would enjoy some support if you can help 😄 

Here is my website: https://mateuswetah.com.br.

If you check my webfinger: https://webfinger.net/lookup/?resource=https%3A%2F%2Fmateuswetah.com.br%2F%40mateuswetah, (which I created manually as suggested [here](https://github.com/Automattic/wordpress-activitypub/issues/188)), you'll see that is right now set to user `admin`. I tried this before with `mateuswetah@mateuswetah.com.br` because I wanted to have any post from any user but had the feeling that this could be the problem so also tried that, since the `admin` user has content in its [blog page](https://mateuswetah.com.br/author/admin/). 

I tried some of the tweaks suggested previously on `.httacess`. No success. I know that Hostinger uses Let's Encrypt, but they didn't provide much more information about it. I can create and edit files in my WordPress folder only. 

BTW if you search for `mateuswetah` on Mastodon you might find [this user](https://mastodon.social/@mateuswetah@social.lab.cultura.gov.br), which I created years ago when I was digging into Fediverse. I would love to have this working via ActivityPub ❤️

**Reply by mateuswetah** [reply]
@pfefferle have you had any chance to look at it?

**Reply by pfefferle** [reply]
I am on vacation, I will check when I am back. ☺️

**Reply by pfefferle** [reply]
@mateuswetah here is a nice tool, to test your blogs profiles: https://browser.pub/mateuswetah@mateuswetah.com.br

there seem to be an issue with the headers for WebFinger: `Unexpected content type in webfinger response: application/octet-stream"`

It seems that your hoster also uses LightSpeed caching, maybe this also helps: https://www.timnolte.com/2023/04/09/fixing-openlitespeed-caching-for-activitypub-on-wordpress/

**Reply by mateuswetah** [reply]
Hey @pfefferle it took me literally one year to take a moment to look at this and things now work well! Yes I had to do some tweaks to overcome the Hostinger caching situation but now I can finally see my profile and my posts in the BrowserPub tool! 🎉 

https://browser.pub/https%3A%2F%2Fmateuswetah.com.br%2F%3Fpost_type%3Dap_outbox%26p%3D2931

However, I can't find my user (`@mateus@mateuswetah.com.br` or `@mateuswetah@mateuswetah.com.br`) in any other federated network. I only have friends that have some Mastodon accounts and they too [could not find me there](https://mastodon.social/search?q=%40mateus%40mateuswetah.com.br). I read about some "cold start" but after publishing a few photos and waiting a few days nothing happens. Can you try to find me on some of your networks? It is curious how every diagnostic that I run via `curl` and the data from the BrowserPub says things are fine but still no success on real networks 😢

**Reply by pfefferle** [reply]
Have you checked back with Hostinger?

**Reply by mateuswetah** [reply]
Yes but since my webfinger and wellknown-host seem to be fine they seem a bit lost on suggestions. Would you know if there's any other test that I could indicate them to do in order to prove that there might be something missing? Maybe a request to some endpoint?

**Comment by bethanye** [comment]
Same problem, also on Dreamhost + Apache + LetsEncrypt. Following the recommendation to allow access to .well-known referenced [here](https://github.com/pfefferle/wordpress-activitypub/discussions/155#:~:text=Apache%20or%20Nginx%3F-,%238%20(comment),-Comment%20options), I had Dreamhost Support add a bypass for .well-known. 

I'm still getting a 400, and `missing "resource" parameter` when I curl the page or visit in a browser.

```
% curl "https://bethanye.geekery.org/.well-known/webfinger?resource=acct:bethanye@bethanye.geekery.org"
missing "resource" parameter%
```

**Reply by davemacdo** [reply]
I am also having the `missing "resource" parameter` problem. Did you ever resolve this?

**Comment by kvibber** [comment]
I put this in my top-level .htaccess and it seems to be working. Also Dreamhost+Apache+Let's Encrypt, but on a VPS, so I'm not sure if that makes a difference.

RewriteRule ^.well-known/(host-meta|webfinger|nodeinfo|x-nodeinfo) /blog/.well-known/$1 [L]

I'm able to find and follow the author account through Mastodon again

**Reply by AndySoc1al** [reply]
Well, that didn't throw an error for me (Dreamhost/Apache/Let's Encrypt non-VPS), but it didn't do anything useful either. The inclusion of /blog/ in your rewrite rule makes me assume that's where WordPress is installed for you. Since I installed WordPress as the root level of my site, I tried removing the /blog portion, and then my site was broken. Ah, well, another approach is apparently needed...

**Reply by timnolte** [reply]
@AndySoc1al I tested and confirmed that your actual JSON endpoint is working so a rewrite for that should work.

https://andysocial.com/wp-json/activitypub/1.0/webfinger?resource=acct:andysocial@andysocial.com

However, the odd thing that I see is that the author that is being returned in the endpoint is `gary` not `andysocial`. Anyways, the rewrite is what you need but perhaps your hosting won't allow the override.

**Reply by AndySoc1al** [reply]
@timnolte Yeah, I don't have any idea what to do with this information. I put that rewrite rule in multiple places in the .htaccess file for my site, and it makes no difference. I don't know how you even "tested" to find that the JSON endpoint is working, since I just get "error" on any attempt to access the webfinger URL, with either a normal browser or curl on the command line. I'm utterly lost on how to make this do anything useful. Thanks for attempting to help.

**Reply by timnolte** [reply]
Have you reached out to Dreamhost for any support? You should only have 1 `.htaccess` file in place, the way they work is by searching current then parent directories. You should be able to place one in the `.well-known` directory. That will also prevent you from breaking other parts of your site. You can use this site(https://htaccess.madewithlove.com/) to test out your htaccess rules as well.

**Reply by timnolte** [reply]
Andy, now that I look again at what kvibber posted that isn't what you are going to need if it's shared hosting. Kvibber said they are on a VPS which means let's encrypt is probably not installing into the WordPress directory. Going to run a test to get something that should work for you.

**Reply by AndySoc1al** [reply]
@timnolte I do only have the one .htaccess file. When I wrote "multiple places" it was in different blocks of the file, either at the top or bottom or whatever. Gave the same result as not having the rule at all. Dreamhost doesn't support the plugin, so they're not much help.

**Reply by timnolte** [reply]
Andy, it might help if you have an existing file that you are editing to see a copy of that in order to direct you. However, to make it easy I would just create a new `.htaccess` file in the `.well-known` directory. Add the following to that file as I've ran this through the tester and should work.

```
RewriteRule ^.well-known/(host-meta|webfinger|nodeinfo|x-nodeinfo)$ /wp-json/activitypub/1.0/$1 [L]
```

**Reply by AndySoc1al** [reply]
Well, look at that - there's an .htaccess file in the webfinger directory, put there apparently by Let's Encrypt. Here's the contents:

>Order allow,deny
>Allow from all
>
>RewriteCond %{REQUEST_URI} ^/[.]well-known/acme-challenge/[a-zA-Z0-9_-]+$
>RewriteRule .* - [L]
>
>RewriteRule .* - [F]
>RewriteRule ^.well-known/(host-meta|webfinger|nodeinfo|x-nodeinfo)$ /wp-json/activitypub/1.0/$1 [L]

After adding your code at the end, I deleted the cache in wp-admin, and still get the same error in Site Health. I'm reluctant to nuke the ACME Challenge code, since I kind of want to keep the SSL cert functional.

I also put the webfinger line just after the Allow line. This seems like it should work, but does not seem to. I'm going to temporarily disable the cache and see if that works.
**Update** Tried disabling the caching plugin, and made no difference. According to the htaccess checker, putting the activitypub rewrite rule at the beginning leads to the wp-json URL, but actually going to the webfinger URL still just throws an error. *sigh*

**Reply by timnolte** [reply]
I would put the line I provided before the acme-challenge line.

**Reply by AndySoc1al** [reply]
> I would put the line I provided before the acme-challenge line.

Did that. No difference. Put it in the root .htaccess as the first line - no difference. I can get to the wp-json virtual URL (virtual because going into SSH on the host shows there is no actual directory named wp-json), but the redirect fails to do anything useful. I'll put in a detailed customer service query to Dreamhost, but I'm guessing they will suggest upgrading to a dedicated IP.

Thanks for all your help, @timnolte

**Reply by AndySoc1al** [reply]
OK, I tried a hack that worked for someone (not on Dreamhost) to force things: copy the contents of that json file that is generated in the wp-json virtual URL and paste it into a text file named webfinger the ./well-known directory. Still the same empty "Error" reply. Decided to be particularly nuts, and created a directory named webfinger, put an HTML file in it and tried to get to **that** URL - same empty "Error" message.

So, yeah, there's something particularly thorough that Dreamhost must be doing with the .well-known folder.

**Reply by huckabarry** [reply]
I tried just about everything as well on BlueHost and finally just gave up on Wordpress and got a Micro.blog this week.

**Comment by FoxLee** [comment]
I have some really weird info to add to this issue. I have the plugin working (on my Dreamhost VPS with LetsEncrypt) on my www domain (https://www.invincible.ink -> `@Updates@www.invincible.ink`), but it's failing with a second WP install on a subdomain (https://press.invincible.ink -> `@Talen@press.invincible.ink`).

The setup on the www domain has an interesting quirk: The wordpress install is actually a couple of subfolders deep compared to the domain root, but I used a workaround to force the www as the wordpress URL for the sake of having it look nice. This involved copy/pasting the normal (core) WordPress htaccess snippet into the htaccess file in the domain root, then adding an index.php that loads the Wordpress header—and thus the entire software:

```
define('WP_USE_THEMES', true);
require( dirname( __FILE__ ) . '/shared/blog/wp-blog-header.php' );
```

As weird as this is, it's always worked like a charm and has run the blog site correctly without getting in the way of a dozen other subdomains and subfolders. More importantly, that's just about the only meaningful difference between the two sites I'm trying to get working. And it does mean that the actual folder where the blog is located doesn't contain .well-known/.htaccess. Could that be why it's behaving there, in spite of known Dreamhost issues, and not on the second site?

**Reply by FoxLee** [reply]
Further to this, I now have it working on several other sites on subdomains on this account—https://escarnum.invincible.ink and https://squarefireballs.invincible.ink—and on another TLD on this account with a conventional setup, https://www.ex-manus.com. They all seem to have weird webfinger lookup results, but I can find and follow their accounts from Mastodon with no issues. I am utterly at a loss as to why only the press.invincible.ink subdomain refuses to work :\

**Comment by caranha** [comment]
Hello! I'm also having issues getting this plugin to work. 

I'm on Wordpress 6.1.1, installed on a linode VPS. I do have let's encrypt. 

After installing, got the "not valid json" message on site health, and went digging here for answers. Tested my website on webfinger.net/lookup, and got that my webfinger is 404'ing.

Saw that several answers involved messing with .htaccess, and checked in my server: my wordpress install directory does not seem to have any explicit .htaccess file (it is a pretty vanilla install, I did almost no hacking other than soft-linking some other directories for static files), and no .well-known directory either. So I'm finding all the "add this to your .htaccess file" suggestions in this thread not very useful :-( 

Any suggestions of things that I should try are welcome!

Extra info:
- Website: https://claus.castelodelego.org
- User: https://claus.castelodelego.org/?author=1
- Site health error message:
```
Your WebFinger endpoint https://claus.castelodelego.org/.well-known/webfinger
?resource=acct:caranha@claus.castelodelego.org does not return valid JSON for application/jrd+json.
```
- Plugins: Webfinger, ActivityPub, Askimet, NextGen Gallery, Classic Editor

**Reply by caranha** [reply]
Note: This workaround worked for me (manually creating .well-known/webfinger)

https://github.com/pfefferle/wordpress-activitypub/issues/188

That said, I'm still interested in finding a "proper" solution to this problem, since it seems to be pretty common. I'd be happy to do stuff to help diagnose it!

**Reply by pfefferle** [reply]
@caranha you have to enable "pretty permalinks" in the WordPress settings: `/wp-admin/options-permalink.php`

**Reply by caranha** [reply]
@pfefferle Thank you! That helped me fix the problem.

I tried to enable "pretty permalinks", and my posts all became 404's. 

Turns out that my wordpress site was not configured on the apache side to use modRewrite. After I added 

```
<Directory>
   # ... other configuration ...
   AllowOverwrite all
</Directory>
```

To the site configuration, both permalinks and the ActivityPub plugin started working normally.

**Reply by caranha** [reply]
Note: My "site-health" still shows the same error message, even though everything seems to work and show valid data now...

**Reply by pfefferle** [reply]
I am glad that it works now! It might be possible that site-help has some kind of caching. I would assume that the error will disappear soon.
